feat(schematics): add ai-config schematic for generating AI tool configurations

fahrigedik opened the pull request:
Resolves #23871 

# AI Configuration Schematic

## Overview

The `ai-config` schematic automatically generates AI tool configuration files for ABP Framework Angular projects. These configurations help AI tools (Claude, Copilot, Cursor, etc.) follow Angular and ABP best practices.

---

## Quick Start

```bash
# Single tool
ng g @abp/ng.schematics:ai-config --tool=cursor

# Multiple tools
ng g @abp/ng.schematics:ai-config --tool=claude --tool=copilot --tool=cursor
```

---

## Supported AI Tools

| Tool | Configuration File |
|------|-------------------|
| **Claude** | `.claude/CLAUDE.md` |
| **GitHub Copilot** | `.github/copilot-instructions.md` |
| **Cursor** | `.cursor/rules/cursor.mdc` |
| **Google Gemini** | `.gemini/GEMINI.md` |
| **Junie AI** | `.junie/guidelines.md` |
| **Windsurf** | `.windsurf/rules/guidelines.md` |

---

## Command Options

### `--tool` (required)
AI tools to generate configuration for.

**Values:** `claude`, `copilot`, `cursor`, `gemini`, `junie`, `windsurf`

```bash
ng g @abp/ng.schematics:ai-config --tool=cursor --tool=claude
```

### `--targetProject` (optional)
Target project name. Defaults to workspace root.

```bash
ng g @abp/ng.schematics:ai-config --tool=cursor --targetProject=my-app
```

### `--overwrite` (optional)
Overwrite existing files. Default: `false`

```bash
ng g @abp/ng.schematics:ai-config --tool=cursor --overwrite
```

### `--dry-run` (optional)
Preview changes without creating files.

```bash
ng g @abp/ng.schematics:ai-config --tool=cursor --dry-run
```

---

## Usage Examples

### First-Time Setup
```bash
ng g @abp/ng.schematics:ai-config --tool=cursor --tool=copilot
```

### Update Existing Configuration
```bash
ng g @abp/ng.schematics:ai-config --tool=cursor --overwrite
```

### Team Setup (package.json)
```json
{
  "scripts": {
    "setup:ai": "ng g @abp/ng.schematics:ai-config --tool=claude --tool=copilot --tool=cursor --overwrite"
  }
}
```

---

## Generated Configuration Includes

All configuration files include guidelines for:

- ✅ Angular best practices (OnPush, RxJS, lifecycle hooks)
- ✅ ABP Framework integration (localization, permissions, API proxies)
- ✅ TypeScript strict mode and proper typing
- ✅ State management with NGXS
- ✅ Testing strategies (Jest)
- ✅ Security and accessibility standards
- ✅ Performance optimization

---

## Troubleshooting

### Configuration Not Recognized
Restart your IDE/editor to load the new configuration.

### File Already Exists
Use `--overwrite` flag to replace existing files:
```bash
ng g @abp/ng.schematics:ai-config --tool=cursor --overwrite
```

### Schematic Not Found
Update to the latest version:
```bash
npm update @abp/ng.schematics
```

---

## FAQ

**Q: Should I commit these files to Git?**  
A: Yes! This ensures team consistency and helps new members.

**Q: Can I customize the generated files?**  
A: Yes, but use `--overwrite` carefully as it will replace customizations.

**Q: Does this affect build or runtime?**  
A: No. These are development-only files not included in the bundle.

**Q: Can I use multiple AI tools?**  
A: Yes! Generate configurations for all tools you use.

**Q: How often should I update?**  
A: After major ABP/Angular updates or when team standards change.

---


How to test it?
1. you have to yarn install in `abp\npm\ng-packs`
2. you have to run command `npm run build:schematics` on `npm\ng-packs\packages`
3. you have to run command `npm link` on  `npm\ng-packs\dist\packages\schematics`
4. you must create new abp project and enter terminal in angular folder
5. you have to run command ` npm link @abp/ng.schematics --legacy-peer-deps` on angular folder in new project
6. you have to test it ` ng g @abp/ng.schematics:ai-config --tool=cursor, claude` on angular folder

fahrigedik committed:
Add AI config schematic for Angular projects

Introduces a new 'ai-config' schematic that generates configuration files for various AI tools (Claude, Copilot, Cursor, Gemini, Junie, Windsurf) to guide code generation and best practices in Angular and ABP Framework projects. Includes schematic implementation, schema, templates for each tool, and updates to collection.json and build scripts.

fahrigedik committed:
Unify ABP + Angular full-stack guidelines across AIs

Replaces individual AI-specific Angular and ABP Framework rules with a single, comprehensive set of full-stack development guidelines. The new rules emphasize modular architecture, strict typing, best practices for both .NET (ABP) and Angular, and ensure consistency, maintainability, and performance across backend and frontend. This change standardizes expectations and coding conventions for all AI assistants in the project.

sumeyyeKurtulus reviewed (commented):
Hello @fahrigedik thank you for putting an effort to add such important feature 🙌🏻 
We could consider the points I mention below before proceeding the tests:

- [x] We could introduce the customization points briefly as they are widely used and would make the prompt special to ABP Angular
https://abp.io/docs/latest/framework/ui/angular/customization-user-interface

- [x] It seems working without any parameter set, however it gives such an error
<img width="503" height="146" alt="Screenshot 2025-10-09 at 14 23 06" src="https://github.com/user-attachments/assets/d8cd5658-8b65-4a62-b17e-b2ff921de7ce" />
<img width="915" height="183" alt="Screenshot 2025-10-09 at 14 23 36" src="https://github.com/user-attachments/assets/a2077fe7-0d8c-4cc2-95c4-e1f93125f902" />


- [x] That would be the best if we could make the users select multiple options by using such commands. It works fine if we provide them one by one, however we can enhance the usage to be accepting an array if possible.
<img width="832" height="348" alt="Screenshot 2025-10-09 at 14 26 35" src="https://github.com/user-attachments/assets/0496dbf1-4dea-4484-b947-4d568d16bbdc" />

fahrigedik committed:
Refactor ai-config to use comma-separated tool string

Updated the ai-config schematic to accept a comma-separated string for the 'tool' option instead of an array. Adjusted validation, parsing, and schema prompts accordingly for improved CLI usability and flexibility.

fahrigedik committed:
Add ABP Angular component replacement documentation

Expanded guidelines in multiple AI config instruction files to include details and examples for using ABP Angular's ReplaceableComponentsService. This addition covers key features, usage, important notes, and links to full documentation for customizing and replacing default components, layouts, and UI elements.

sumeyyeKurtulus reviewed (commented):
Thank you for checking the points and making enhancements along with that. I just can comment on how we accept the tool list parameter.

It works fine once we do not give any space. However,  it is better if we trim the spaces while executing the command since the user may specify a space as I did.

![Image 17 10 2025 at 13 32](https://github.com/user-attachments/assets/e0088142-e008-4846-bbb9-0620a22c5256)

fahrigedik committed:
Merge branch 'dev' into issue-23871

fahrigedik committed:
Trim targetProject input before lookup

Ensures that leading and trailing whitespace is removed from the targetProject option before attempting to retrieve the project from the workspace. This prevents errors when users accidentally include extra spaces in the project name.

sumeyyeKurtulus reviewed (commented):
> It works fine once we do not give any space. However, it is better if we trim the spaces while executing the command since the user may specify a space as I did.


Hello again @fahrigedik thank you for adding the trim, but I also mean that such parameter `tools=cursor, gemini` should be parsed into `cursor` and `gemini`. So we need to remove all the possible spaces.

Since I have merged this PR beforehand https://github.com/abpframework/abp/pull/24026, that would be the best if you could resolve the conflicts by getting a merge from dev.

fahrigedik committed:
Merge branch 'dev' into issue-23871

fahrigedik committed:
Simplify tool string splitting in ai-config command

Refactored the splitting of the 'tool' option to use a single regex for whitespace and commas, removing the need for an explicit trim(). This streamlines parsing and improves code clarity.

sumeyyeKurtulus committed:
update: ai-tool usage examples

sumeyyeKurtulus reviewed (approved):
Hello @fahrigedik thank you for your effort on building this. I have just updated the usage examples and it is ready to be tested @yagmurcelk
